pkg/frontend/transport/handler.go

```diff
@@ -119,7 +126,36 @@ func NewHandler(cfg HandlerConfig, roundTripper http.RoundTripper, log log.Logge
 	return h
 }
 
+// Stop makes f enter stopped mode and wait on in-flight requests.
+func (f *Handler) Stop() {
+	f.mtx.Lock()
+	f.stopped = true
+
+	level.Debug(f.log).Log("msg", "waiting on in-flight requests", "requests", f.inflightRequests)
```

Comment:
I would make both logs here an Info. Reason is that it's useful to quickly understand why query-frontend may take some time to terminate.

Reply:
Makes sense @pracucci - I can make a follow-up PR.

Reply:
[PR](https://github.com/grafana/mimir/pull/4170).